### PR TITLE
[versioning] EOS-24954: Delete specific version

### PR DESCRIPTION
**Pull request by ajay-paratmandali**
# Problem Statement
- Delete specific version of object

# Design
-  https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/773062787/DeleteObject+LLD

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [x] Interface change (if any) are documented
- [x] Side effects on other features (deployment/upgrade)
- [x] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: server/s3_delete_object_action.h [hunk 1]
```diff
@@ -54,6 +54,7 @@ class S3DeleteObjectAction : public S3ObjectAction {
 
   // delete marker metadata
   std::shared_ptr<S3ObjectMetadata> delete_marker_metadata;
+  bool is_object_delete_marker;
```
Comment (keithpine) [inline server/s3_delete_object_action.h:57]: Is it necessary to keep this copy? It seems anywhere you are using this, `object_metadata->is_delete_marker()` would work just fine?

**Review by keithpine — APPROVED** [review]
See comment.

File: server/s3_delete_object_action.h (hunk 1 above)
Comment (ajay-paratmandali) [reply server/s3_delete_object_action.h:57]: it is needed 
after object metadata deleted we need to check in cleanup that is it delete marker or not
and during cleanup we can not use object_metadata->is_delete_marker()

**Review by sjain09 — APPROVED** [review]
looks good.

File: server/s3_delete_object_action.cc [hunk 2]
```diff
@@ -354,6 +365,14 @@ void S3DeleteObjectAction::add_oids_to_probable_dead_oid_list() {
   next();
 }
 
+void S3DeleteObjectAction::delete_version_metadata() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  object_metadata->remove_version_metadata(
```
Comment (indrajitzagade) [inline server/s3_delete_object_action.cc:370]: I hope this deletes the metadata from the object version list index table.

**Review by indrajitzagade — APPROVED** [review]

File: server/s3_delete_object_action.cc (hunk 2 above)
Comment (ajay-paratmandali) [reply server/s3_delete_object_action.cc:370]: Yes object_metadata load by parent class, if version id provided then it load from version index else object index

**Comment by ajay-paratmandali** [comment]
PreMerge: http://eos-jenkins.colo.seagate.com/job/S3server/job/S3Server-PreMerge-Test/855/

File: ut/s3_delete_object_action_test.cc [hunk 3]
```diff
@@ -147,6 +147,7 @@ class S3DeleteObjectActionTest : public testing::Test {
   int call_count_one;
   int layout_id;
   std::string bucket_name, object_name;
+  bool is_object_delete_marker = false;
```
Comment (sanalpk) [inline ut/s3_delete_object_action_test.cc:150]: Are you using this variable?

File: server/s3_delete_object_action.h [hunk 4]
```diff
@@ -138,8 +141,10 @@ class S3DeleteObjectAction : public S3ObjectAction {
...
   FRIEND_TEST(S3DeleteObjectActionTest, DeleteObjectsDelayedEnabled);
   FRIEND_TEST(S3DeleteObjectActionTest, DeleteObjectsDelayedDisabled);
-  FRIEND_TEST(S3DeleteObjectActionTest, DeleteHandlerInEnabledState);
+  FRIEND_TEST(S3DeleteObjectActionTest, DeleteHandlerCreateDeleteMarker);
   FRIEND_TEST(S3DeleteObjectActionTest, MetadataHandlerInEnabledState);
+  FRIEND_TEST(S3DeleteObjectActionTest, DeleteHandlerDeleteDeleteMarker);
+  FRIEND_TEST(S3DeleteObjectActionTest, DeleteVersionMetadata);
 };
 
 #endif
```
Comment (sanalpk) [inline server/s3_delete_object_action.h:150]: Any ST cases required for this changes?

File: ut/s3_delete_object_action_test.cc (hunk 3 above)
Comment (ajay-paratmandali) [reply ut/s3_delete_object_action_test.cc:150]: is_object_delete_marker 
during cleanup if object_metadata is delete marker 
then we need to mark operation done

but during cleanup metadata is already deleted 
so variable used to to check if it is delete marker

File: server/s3_delete_object_action.h (hunk 4 above)
Comment (ajay-paratmandali) [reply server/s3_delete_object_action.h:150]: ST will be added after delete first and last index available
as we can not delete bucket
added in separate ticket

**Review by sanalpk — APPROVED** [review]

**Comment by indrajitzagade** [comment]
Build job: http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/cortx-all-image-custom-ci/1583/ (Passed)
Deployment job: http://eos-jenkins.colo.seagate.com/job/Cortx-kubernetes/job/deploy-cortx-k8-cluster/603/ (Passed)

Running the Pre-merge job

**Pull request merged**